README.md
# Retour plugin for Craft CMS

Retour allows you to intelligently redirect legacy URLs, so that you don't lose SEO value when rebuilding & restructuring a website.

![Screenshot](resources/screenshots/retour01.png)

## Installation

To install Retour, follow these steps:

1. Download & unzip the file and place the `retour` directory into your `craft/plugins` directory
2.  -OR- do a `git clone https://github.com/nystudio107/retour.git` directly into your `craft/plugins` folder.  You can then update it with `git pull`
3.  -OR- install with Composer via `composer require nystudio107/retour`
4. Install plugin in the Craft Control Panel under Settings > Plugins
5. The plugin folder should be named `retour` for Craft to see it.  GitHub recently started appending `-master` (the branch name) to the name of the folder for zip file downloads.

Retour works on Craft 2.4.x, Craft 2.5.x, and Craft 2.6.x.

## Retour Overview

Retour allows you to intelligently redirect legacy URLs, so that you don't lose SEO value when rebuilding & restructuring a website.

In addition to supporting traditional exact and RegEx matching of URL patterns, Retour also has a Retour Redirect FieldType that you can add to your entries. This allows you to have dynamic entry redirects that have access to the data in your entries when matching URL patterns.

Retour will also automatically create a redirect for you if you change an entry's slug, or move an entry around in a Structure.

Retour is written to be performant. There is no impact on your website's performance until a 404 exception happens; and even then the resulting matching happens with minimal impact.

Don't just rebuild a website. Transition it with Retour.

## Why Use a Plugin for Redirects?

If you have just a few static redirects, then your best bet is to put them in your `.htaccess` file, or better yet, in your `.conf` file for your virtual host.  However, there are a number of cases where using a plugin to handle it is a **better** solution:

1. If you have a large number of redirects, it will slow down every single request your web server handles unnecessarily if they are in `.htaccess` or `.conf`
2. Often the URL patterns from the legacy website do not match the new website URLs in a deterministic way, which makes creating redirects difficult
3. Sometimes you don't have access to the server config files, or you want to give your client the ability to manage redirects easily

Retour solves these problems:

1. Retour only attempts to do a redirect after the web server has already thrown a 404 exception.  Once a redirect mapping is successfully determined, it also caches the result for speedy resolution of the next redirect request.
2. Retour also gives you the ability to do Dynamic Entry Redirects that allow you to import a piece of legacy data into your entries to use as a key for determining the new URL mapping.  In this way, utterly dissimilar URLs can be mapped for redirection effectively.
3. It provides an easy to use GUI that the client can use from Craft's AdminCP, and keeps statistics on the 404 hits (and misses)

### A Word about .htaccess

People using the Apache webserver are familiar with the `.htaccess` file, and may even be using it for redirects.  It's very likely that you should not be using `.htaccess` at all; instead you should disable `.htaccess` via `AllowOverride none` and make your configuration changes in your webserver configuration files.  From [Apache HTTP Server Tutorial: .htaccess files](https://httpd.apache.org/docs/current/howto/htaccess.html)

    There are two main reasons to avoid the use of .htaccess files.

    The first of these is performance. When AllowOverride is set to allow the
    use of .htaccess files, httpd will look in every directory for .htaccess
    files. Thus, permitting .htaccess files causes a performance hit, whether or
    not you actually even use them! Also, the .htaccess file is loaded every
    time a document is requested.

    Further note that httpd must look for .htaccess files in all higher-level
    directories, in order to have a full complement of directives that it must
    apply. (See section on how directives are applied.) Thus, if a file is
    requested out of a directory /www/htdocs/example, httpd must look for the
    following files:

    /.htaccess
    /www/.htaccess
    /www/htdocs/.htaccess
    /www/htdocs/example/.htaccess

    And so, for each file access out of that directory, there are 4 additional
    file-system accesses, even if none of those files are present. (Note that
    this would only be the case if .htaccess files were enabled for /, which is
    not usually the case.)

    In the case of RewriteRule directives, in .htaccess context these regular
    expressions must be re-compiled with every request to the directory, whereas
    in main server configuration context they are compiled once and cached.
    Additionally, the rules themselves are more complicated, as one must work
    around the restrictions that come with per-directory context and
    mod_rewrite. Consult the Rewrite Guide for more detail on this subject.

As you can see, avoiding the use of `.htaccess` completely is best if at all possible, and especially avoid it for `RewriteRule` directives, such as 404 rewrites.

## Dynamic Entry Redirects

Retour implements a Retour Redirect FieldType that you can add to your Entry Types.  Retour will look for 404 (Not Found) URLs that match the Legacy URL Pattern, and redirect them to this entry's URL.

You also get the context of the `entry` that you can use when matching legacy URLs; so if you've imported a field called `recipeid` into your new website, you can the Retour Redirect FieldType look for it in your Legacy URL Pattern, e.g.: `/old-recipes/{recipeid}`

This allows you to key off of a piece of legacy data that was imported, for the cases when the new URL patterns don't look anything like the Legacy URL Patterns, or follow any pattern that RegEx is useful for matching.

![Screenshot](resources/screenshots/retour02.png)

### Creating a Retour Redirect Field

Create a Retour Redirect field as you would any other field; then set the default values for it.  For new entries, it will default to the values entered here, so you can put your matching pattern in once, rather than having to do it for each entry.

* **Default Legacy URL Pattern** - Enter the URL pattern that Retour should match. This matches against the path, the part of the URL after the domain name. You can include tags that output entry properties, such as `{title}` or `{myCustomField}` in the text field below. e.g.: Exact Match: `/recipes/{recipeid}` or RegEx Match: `.*RecipeID={recipeid}$` where `{recipeid}` is a field handle to a field in this entry.
* **Default Pattern Match Type** - What type of matching should be done with the Legacy URL Pattern. Details on RegEx matching can be found at [regexr.com](http://regexr.com). If a plugin provides a custom matching function, you can select it here.
* **Default Redirect Type** - Select whether the redirect should be permanent or temporary.
* **Redirect Changeable** - Whether to allow the user to change the redirect while editing the entry.

### Configuring a Retour Redirect Field


* **Legacy URL Pattern** - Enter the URL pattern that Retour should match. This matches against the path, the part of the URL after the domain name. You can include tags that output entry properties, such as `{title}` or `{myCustomField}` in the text field below. e.g.: Exact Match: `/recipes/{recipeid}` or RegEx Match: `.*RecipeID={recipeid}$` where `{recipeid}` is a field handle to a field in this entry.
* **Pattern Match Type** - What type of matching should be done with the Legacy URL Pattern. Details on RegEx matching can be found at [regexr.com](http://regexr.com). If a plugin provides a custom matching function, you can select it here.
* **Redirect Type** - Select whether the redirect should be permanent or temporary.

**Note:** if you add a Retour Redirect FieldType to an existing Section, or you import data from a foreign source into a Section with a Retour Redirect FieldType, the default values you set for the Retour Redirect FieldType will not be propagated to the entry yet.  To cause that to happen, go to **Settings->Sections** then click on the Section to edit it, and hit **Save**.  This will cause all of the entries in that section to be re-saved, and Retour will fill in the default field values.

## Static Redirects

### Manually Creating Static Redirects

Static Redirects are useful when the Legacy URL Patterns and the new URL patterns are deterministic.  You can create them by clicking on **Retour->Redirects** and then clicking on the **+ New Static Redirect** button.

* **Legacy URL Pattern** - Enter the URL pattern that Retour should match. This matches against the path, the part of the URL after the domain name. e.g.: Exact Match: `/recipes/` or RegEx Match: `.*RecipeID=(.*)`
* **Destination URL** - Enter the destination URL that should be redirected to. This can either be a fully qualified URL or a relative URL. e.g.: Exact Match: `/new-recipes/` or RegEx Match: `/new-recipes/$1`
* **Pattern Match Type** - What type of matching should be done with the Legacy URL Pattern. Details on RegEx matching can be found at [regexr.com](http://regexr.com). If a plugin provides a custom matching function, you can select it here.
* **Redirect Type** - Select whether the redirect should be permanent or temporary.

### Importing an Existing .htaccess file

Retour also allows you to import an existing `.htaccess` file and all of the redirects it contains into Retour by clicking on **Retour->Redirects** and then clicking on the **Import .htaccess File** button.

It will import redirects from `Redirect` and `RedirectMatch` directives in the file.  It will **ignore** `RewriteRule`s because they don't necessarily have a 1:1 mapping, you can have several `RewriteRule`s that are strung together to figure out the final redirect.

It asks your browser to look for only `text` files to upload; if the `.htaccess` file you have isn't a `.txt` file, you can force it to allow you to upload it by choosing **Format: All Files**.

### Renamed Slug Redirects

If you rename an entry's `slug` (and the Section the entry is in has URLs), Retour will automatically create a static redirect for you to keep traffic going to the right place.  It will also automatically create a static redirect if you move an entry around in a Structure.

It will appear listed under the "Static Redirects" section like any other static redirect.

## Retour Statistics

Retour keeps track of every 404 your website receives.  You can view them by clicking on **Retour->Statistics**.  

Only one record is saved per URL Pattern, so the database won't get clogged with a ton of records.


CHANGELOG.md
# Retour Changelog

## 1.0.22 - 2017.10.02
### Changed
* Fixed an issue in the `Retour_StatsModel`
* Exact matches are now checked before RegEx matches for static redirects
* Fixed the documentation that was incorrect about being able to add multiple FieldTypes to a single entry
* Replace 4-byte UTF-8 characters in `Retour_Statistics` before saving them, to avoid db errors

## 1.0.21 - 2017.08.31
### Changed
* Fixed a hardcoded table name in the `referrerUrl` migration

## 1.0.20 - 2017.08.30
### Added
* Retour will automatically trim the `retour_stats` table to the last 10,000 redirects, sorted by date (configurable via `statsStoredLimit` in `config.php`)
* Add the URL to the stats title attribute, for cases where the display is truncated
* Respect `addTrailingSlashesToUrls` in the URLs returned from `getLocalizedrUls()`
* Updated README.md to note that importing `.htaccess` ignores `RewriteRule`
* Added a `alwaysStripQueryString` setting to `config.php` (defaults to `false`)
* Added a `stripQueryStringFromStats` setting to `config.php` (defaults to `true`)

### Changed
* Refactored and cleaned up the code
* Fixed some deprecated meta information in `RetourPlugin.php`
* The `referrerUrl` has a maximum length of 2000 now
* Fixed some typos in `README.md`

## 1.0.19 - 2017.02.10

* [Added] Added a referrer column in the Stats table
* [Added] Added additional logging in `devMode`
* [Improved] No more default value for `redirectSrcUrl` column (could cause SQL exceptions in newer versions of MySQL)
* [Improved] Updated CHANGELOG.md

## 1.0.18 - 2017.01.20

* [Fixed] The `addTrailingSlashesToUrls` is now respected for dynamic entry redirects
* [Improved] Merged pull request 'Fix retourMatch hook'
* [Added] Added a `statsDisplayLimit` setting to `config.php` to control how many stats should be displayed in the AdminCP
* [Improved] Merged pull request 'Limit returned results to template'
* [Improved] Merged pull request 'Allow handling of 404s from { exit } tags encountered while rendering templates'
* [Added] Added a config.php setting `createUriChangeRedirects` so that the URI-change redirects can be disabled on a per-environment basis
* [Improved] Don't redirect to the welcome page if we're being installed via Console command
* [Improved] Moved the changelog to CHANGELOG.md

## 1.0.17 - 2016.08.31

* [Improved] Query strings are now stripped from the incoming URI before redirect detection is done
* [Improved] Updated the README.md

## 1.0.16 - 2016.08.30

* [Fixed] FieldTypes in multi-locale setups that are not translatable are now handled properly
* [Fixed] Fixed missing locale prefix for localized entries in the FieldType
* [Fixed] Fixed an issue where FieldType redirects had an errant / prepended to them
* [Improved] Better importing of `.htaccess` files
* [Improved] Better error handling when importing malformed `.htaccess` files
* [Fixed] Trailing /'s are no longer stripped from URLs added via the `+` icon from the Statistics page
* [Fixed] Fixed an issue that would prevent RegEx's from matching as they should in FieldTypes
* [Improved] Updated the README.md

## 1.0.15 - 2016.07.12

* [Added] Added the ability to import the redirects from a `.htaccess` file into Retour
* [Fixed] Fixed a statics db error with empty referrers
* [Improved] Updated the README.md

## 1.0.14 - 2016.07.10

* [Added] The Statistics and Redirects tables are now dynamically searchable and sortable
* [Fixed] Fixed an issue that caused redirects created via the `+` from Statistics page to not save
* [Improved] Updated the README.md

## 1.0.13 - 2016.07.06

* [Added] Adds support for locales in the automatic redirect that is created when a slug is changed for an entry
* [Improved] Retour will no longer let you save a static redirect with an empty destinationURL
* [Fixed] Fixed a typo in the Retour_StatsModel
* [Improved] Added a rant about `.htaccess` to the docs
* [Improved] Updated the README.md

## 1.0.12 - 2016.07.04

* [Added] If you hover over a 404 File Not Found URL on the Statistics page, you'll now see the last referrer for the 404 URL
* [Added] Added a + button on the Statistics page that lets you quickly add a 404'd URL as a redirect
* [Improved] We now store the destination for redirects in the FieldType as a URI rather than a URL, so that it's more portable across environments
* [Added] Structure entries that have Retour FieldTypes in them now have the destinationURL updated when the structure elements are moved
* [Improved] The widget now handles very long URLs more gracefully
* [Improved] Updated the README.md

## 1.0.11 - 2016.06.21

* [Fixed] Fixed an issue with URLs that have umlauts in them
* [Fixed] Fixed an issue with URLs that are longer than 255 characters for the redirect statistics
* [Improved] Statistics are now limited to the top 1,000 hits
* [Improved] Updated the README.md

## 1.0.10 - 2016.06.15

* [Added] Retour will attempt to prevent redirect loops when saving a new redirect by deleting any existing redirects that have the destUrl as their srcUrl
* [Added] Added a 410 - Gone redirect http code for permanently removed resources
* [Improved] Updated the README.md

## 1.0.9 - 2016.06.04

* [Added] Retour will now automatically create a static redirect for you if you rename an entry's slug
* [Improved] Retour checks to ensure that no two redirects have the same redirectSrcUrl

services/RetourService.php
<?php
/**
 * Retour plugin for Craft CMS
 *
 * Retour Service
 *
 * @author    Andrew Welch
 * @copyright Copyright (c) 2016 nystudio107
 * @link      http://nystudio107.com
 * @package   Retour
 * @since     1.0.0
 */

namespace Craft;

class RetourService extends BaseApplicationComponent
{

    protected $cachedEntryRedirects = null;
    protected $cachedStaticRedirects = null;
    protected $cachedStatistics = null;

    /**
     * @return array All of the statistics
     */
    public function getAllStatistics()
    {

        // Cache it in our class; no need to fetch it more than once
        if (isset($this->cachedStatistics)) {
            return $this->cachedStatistics;
        }

        $result = craft()->db->createCommand()
            ->select('*')
            ->from('retour_stats')
            ->order('hitCount DESC')
            ->limit(craft()->config->get("statsDisplayLimit", "retour"))
            ->queryAll();

        $this->cachedStatistics = $result;

        return $result;
    }

    /**
     * @param int $days The number of days to get
     * @param int $handled
     *
     * @return array Recent statistics
     */
    public function getRecentStatistics($days = 1, $handled = 0)
    {

        $handled = (int)$handled;

        if (!$handled) {
            $handled = 0;
        }
        $result = craft()->db->createCommand()
            ->select('*')
            ->from('retour_stats')
            ->where("hitLastTime >= ( CURDATE() - INTERVAL '$days' DAY )")
            ->andWhere('handledByRetour =' . $handled)
            ->order('hitLastTime DESC')
            ->queryAll();

        return $result;
    }

    /**
     */
    public function clearStatistics()
    {
        $result = craft()->db->createCommand()
            ->truncateTable('retour_stats');

        return $result;
    }

    /**
     * @param  string $url the url to match
     *
     * @return mixed      the redirect array
     */
    public function findRedirectMatch($url)
    {
        $result = null;

        // Check the cache first
        $redirect = $this->getRedirectFromCache($url);
        if ($redirect) {
            $error = $this->incrementRedirectHitCount($redirect);
            $this->saveRedirectToCache($url, $redirect);
            RetourPlugin::log("[cached] " . $redirect['redirectMatchType'] . " result: " . print_r($error, true), LogLevel::Info, false);

            return $redirect;
        }

        // Look up the entry redirects first
        $redirects = null;
        $redirects = $this->getAllEntryRedirects();
        $result = $this->lookupRedirect($url, $redirects);
        if ($result) {
            return $result;
        }

        // Look up the static redirects next
        $redirects = null;
        $redirects = $this->getAllStaticRedirects();
        $result = $this->lookupRedirect($url, $redirects);
        if ($result) {
            return $result;
        }

        return $result;
    }

    /**
     * @param $url
     *
     * @return mixed The redirect
     */
    public function getRedirectFromCache($url)
    {
        $cacheKey = "retour_cache_" . md5($url);
        $result = craft()->cache->get($cacheKey);
        RetourPlugin::log("Cached Redirect hit: " . print_r($result, true), LogLevel::Info, false);

        return $result;
    }

    /**
     * @param  Retour_RedirectsModel The redirect to create
     */
    public function incrementRedirectHitCount(&$redirect)
    {
        if (isset($redirect)) {
            $redirect['hitCount'] = $redirect['hitCount'] + 1;
            $redirect['hitLastTime'] = DateTimeHelper::currentTimeForDb();

            if ($redirect['associatedElementId']) {
                $table = 'retour_redirects';
            } else {
                $table = 'retour_static_redirects';
            }
            $result = craft()->db->createCommand()
                ->update($table, array(
                    'hitCount'    => $redirect['hitCount'],
                    'hitLastTime' => $redirect['hitLastTime'],
                ), 'id=:id', array(':id' => $redirect['id']));
        }
    }

    /**
     * @param  string $url     The input URL
     * @param  mixed  $redirect The redirect
     */
    public function saveRedirectToCache($url, $redirect)
    {
        $cacheKey = "retour_cache_" . md5($url);
        $error = craft()->cache->set($cacheKey, $redirect, 0);
        RetourPlugin::log("Cached Redirect saved: " . print_r($error, true), LogLevel::Info, false);
    }

    /**
     * @return Array All of the entry redirects
     */
    public function getAllEntryRedirects($limit = null)
    {

        // Cache it in our class; no need to fetch it more than once
        if (isset($this->cachedEntryRedirects)) {
            return $this->cachedEntryRedirects;
        }

        $result = craft()->db->createCommand()
            ->select('*')
            ->from('retour_redirects')
            ->order('hitCount DESC');

        if ($limit) {
            $result = $result->limit($limit);
        } else {
            $this->cachedEntryRedirects = $result;
        }

        return $result->queryAll();
    }

    /**
     * @param  string $url       the url to match
     * @param  mixed  $redirects an array of redirects to look through
     *
     * @return mixed      the redirect array
     */
    public function lookupRedirect($url, $redirects)
    {
        $result = null;
        foreach ($redirects as $redirect) {
            $redirectMatchType = isset($redirect['redirectMatchType']) ? $redirect['redirectMatchType'] : null;
            switch ($redirectMatchType) {
                // Do a straight up match
                case "exactmatch":
                    if (strcasecmp($redirect['redirectSrcUrlParsed'], $url) === 0) {
                        $error = $this->incrementRedirectHitCount($redirect);
                        RetourPlugin::log($redirectMatchType . " result: " . print_r($error, true), LogLevel::Info, false);
                        $this->saveRedirectToCache($url, $redirect);

                        return $redirect;
                    }
                    break;

                // Do a regex match
                case "regexmatch":
                    $matchRegEx = "`" . $redirect['redirectSrcUrlParsed'] . "`i";
                    if (preg_match($matchRegEx, $url) === 1) {
                        $error = $this->incrementRedirectHitCount($redirect);
                        RetourPlugin::log($redirectMatchType . " result: " . print_r($error, true), LogLevel::Info, false);

                        // If we're not associated with an EntryID, handle capture group replacement
                        if ($redirect['associatedElementId'] == 0) {
                            $redirect['redirectDestUrl'] = preg_replace($matchRegEx, $redirect['redirectDestUrl'], $url);
                        }
                        $this->saveRedirectToCache($url, $redirect);

                        return $redirect;
                    }
                    break;

                // Otherwise try to look up a plugin's method by and call it for the match
                default:
                    $plugin = $redirectMatchType ? craft()->plugins->getPlugin($redirectMatchType) : null;
                    if ($plugin) {
                        if (method_exists($plugin, "retourMatch")) {
                            $args = array(
                                array(
                                    'redirect' => &$redirect,
                                ),
                            );
                            $result = call_user_func_array(array($plugin, "retourMatch"), $args);
                            if ($result) {
                                $error = $this->incrementRedirectHitCount($redirect);
                                RetourPlugin::log($redirectMatchType . " result: " . print_r($error, true), LogLevel::Info, false);
                                $this->saveRedirectToCache($url, $redirect);

                                return $redirect;
                            }
                        }
                    }
                    break;
            }
        }
        RetourPlugin::log("Not handled: " . $url, LogLevel::Info, false);

        return $result;
    }

    /**
     * @return array All of the static redirects
     */
    public function getAllStaticRedirects($limit = null)
    {

        // Cache it in our class; no need to fetch it more than once
        if (isset($this->cachedStaticRedirects)) {
            return $this->cachedStaticRedirects;
        }

        $result = craft()->db->createCommand()
            ->select('*')
            ->from('retour_static_redirects')
            ->order('redirectMatchType ASC, hitCount DESC');

        if ($limit) {
            $result = $result->limit($limit);
        } else {
            $this->cachedStaticRedirects = $result;
        }

        return $result->queryAll();
    }

    /**
     * @param $record
     *
     * @return bool|int|string
     */
    public function saveStaticRedirect($record)
    {
        $error = "";

        if (isset($record)) {
            if (($record->redirectSrcUrl == "") || ($record->redirectDestUrl == "")) {
                $id = $record->id;
                $affectedRows = craft()->db->createCommand()->delete('retour_static_redirects', array(
                    'id' => $id,
                ));

                RetourPlugin::log("Deleted Redirected: " . $id, LogLevel::Info, false);
                $error = craft()->cache->flush();
                RetourPlugin::log("Cache flushed: " . print_r($error, true), LogLevel::Info, false);
                $error = -1;
            } else {
                if ($record->save()) {
                    $error = craft()->cache->flush();
                    RetourPlugin::log("Cache flushed: " . print_r($error, true), LogLevel::Info, false);
                    craft()->userSession->setNotice(Craft::t('Retour Redirect saved.'));
                    $error = "";

                    // To prevent redirect loops, see if any static redirects have our destUrl as their srcUrl
                    $redir = $this->getRedirectByRedirectSrcUrl($record->redirectDestUrl, $record->locale);
                    if ($redir) {
                        $id = $redir->id;
                        $affectedRows = craft()->db->createCommand()->delete('retour_static_redirects', array(
                            'id' => $id,
                        ));
                    }
                } else {
                    $error = $record->getErrors();
                    RetourPlugin::log(print_r($error, true), LogLevel::Info, false);
                    craft()->userSession->setError(Craft::t('Couldn’t save Retour Redirect.'));
                }
            }
        }

        return $error;
    }

    /**
     * @param  string $srcUrl the redirect's redirectSrcUrl
     * @param  string $locale The locale
     *
     * @return Mixed The resulting Redirect
     */
    public function getRedirectByRedirectSrcUrl($srcUrl, $locale)
    {
        $result = Retour_RedirectsRecord::model()->findByAttributes(array('redirectSrcUrlParsed' => $srcUrl, 'locale' => $locale));

        return $result;
    }

    /**
     * @param      $url The 404 url
     * @param bool $handled
     */
    public function incrementStatistics($url, $handled = false)
    {

        $handled = (int)$handled;
        $url = substr($url, 0, 255);
        $referrer = craft()->request->getUrlReferrer();
        if (is_null($referrer)) {
            $referrer = "";
        }

        // Strip the query string if `stripQueryStringFromStats` is set
        if (craft()->config->get("stripQueryStringFromStats", "retour")) {
            $url = UrlHelper::stripQueryString($url);
        }

        // Make sure the referrerUrl does not exceed the max length of its table column.
        $attrConfigs = Retour_RedirectsRecord::model()->getAttributeConfigs();
        $maxLength = isset($attrConfigs['referrerUrl']['maxLength']) ? $attrConfigs['referrerUrl']['maxLength'] : 255;
        $trimMarker = '...';
        $referrer = mb_strimwidth($referrer, 0, ($maxLength-strlen($trimMarker)), $trimMarker, craft()->charset);

        // See if a stats record exists already
        $result = craft()->db->createCommand()
            ->select('*')
            ->from('retour_stats')
            ->where('redirectSrcUrl =' . craft()->db->quoteValue($url))
            ->queryAll();

        if (empty($result)) {
            $stats = new Retour_StatsRecord;
            $stats->redirectSrcUrl = StringHelper::encodeMb4($url);
            $stats->referrerUrl = StringHelper::encodeMb4($referrer);
            $stats->hitCount = 1;
            $stats->hitLastTime = DateTimeHelper::currentUTCDateTime();
            $stats->handledByRetour = $handled;
            $stats->save();
        } else {
            // Update the stats table
            foreach ($result as $stat) {
                $stat['hitCount'] = $stat['hitCount'] + 1;
                $stat['hitLastTime'] = DateTimeHelper::currentTimeForDb();
                $stat['referrerUrl'] = $referrer;

                $result = craft()->db->createCommand()
                    ->update('retour_stats', array(
                        'hitCount'        => $stat['hitCount'],
                        'hitLastTime'     => $stat['hitLastTime'],
                        'handledByRetour' => $handled,
                        'referrerUrl'     => $stat['referrerUrl'],
                    ), 'id=:id', array(':id' => $stat['id']));
            }
        }

        // After incrementing a statistic, trim the retour_stats db table
        $this->trimStatistics();
    }

    /**
     * Trim the retour_stats db table based on the statsStoredLimit config.php
     * setting
     *
     * @return void
     */
    public function trimStatistics()
    {
        $affectedRows = 0;
        $table = craft()->db->addTablePrefix('retour_stats');
        $quotedTable = craft()->db->quoteTableName($table);
        $limit = craft()->config->get("statsStoredLimit", "retour");

        // As per https://stackoverflow.com/questions/578867/sql-query-delete-all-records-from-the-table-except-latest-n
        if (!empty($limit) && $limit) {
            $affectedRows = craft()->db->createCommand("
            DELETE FROM $quotedTable
            WHERE id NOT IN (
              SELECT id
              FROM (
                SELECT id
                FROM $quotedTable
                ORDER BY hitLastTime DESC
                LIMIT $limit
              ) foo
            )
        ")->execute();
            RetourPlugin::log("Trimmed " . $affectedRows . " from retour_stats table", LogLevel::Info, false);
        }
    }

    /**
     * @param  int $id The redirect's id
     *
     * @return Mixed The resulting Redirect
     */
    public function getRedirectById($id)
    {
        $result = Retour_StaticRedirectsRecord::model()->findByAttributes(array('id' => $id));

        return $result;
    }

    /**
     * @param  Retour_RedirectsModel The redirect to save
     */
    public function saveRedirect($redirectsModel)
    {
        if (isset($redirectsModel)) {
            $result = $this->getRedirectByElementId($redirectsModel->associatedElementId, $redirectsModel->locale);
            if ($result) {
                $result->setAttributes($redirectsModel->getAttributes(), false);
                $error = $result->save();
            } else {
                $error = $this->createRedirect($redirectsModel);
            }
            RetourPlugin::log(print_r($error, true), LogLevel::Info, false);
        }
    }

    /**
     * @param  int    $elementId The associated elementId
     * @param  string $locale    The locale
     *
     * @return Mixed The resulting Redirect
     */
    public function getRedirectByElementId($elementId, $locale)
    {
        $result = Retour_RedirectsRecord::model()->findByAttributes(array('associatedElementId' => $elementId, 'locale' => $locale));

        return $result;
    }

    /**
     * @param  Retour_RedirectsModel The redirect to create
     */
    public function createRedirect($redirectsModel)
    {
        if (isset($redirectsModel)) {
            // Don't try to create a redirect if one already exists for the redirectSrcUrlParsed, or if empty
            if ($redirectsModel->redirectSrcUrlParsed && !$this->getRedirectByRedirectSrcUrl($redirectsModel->redirectSrcUrlParsed, $redirectsModel->locale)) {
                $result = new Retour_RedirectsRecord;
                $result->setAttributes($redirectsModel->getAttributes(), false);
                $result->save();
                $error = $result->getErrors();
                RetourPlugin::log(print_r($error, true), LogLevel::Info, false);
            }
        }
    }

    /**
     * @param  int    $elementId The associated elementId
     * @param  string $locale    The locale
     */
    public function deleteRedirectByElementId($elementId, $locale)
    {
        $result = $this->getRedirectByElementId($elementId, $locale);
        if ($result) {
            $result->delete();
        }
    }

    /**
     * Returns a list of localized URIs for the passed in element
     *
     * @param null $element
     *
     * @return array an array of paths
     */
    public function getLocalizedUris($element = null)
    {
        $localizedUris = array();
        if ($element) {
            if (craft()->isLocalized()) {
                $unsortedLocalizedUris = array();
                $_rows = craft()->db->createCommand()
                    ->select('locale')
                    ->addSelect('uri')
                    ->from('elements_i18n')
                    ->where(array('elementId' => $element->id, 'enabled' => 1))
                    ->queryAll();

                foreach ($_rows as $row) {
                    $path = ($row['uri'] == '__home__') ? '' : $row['uri'];
                    $url = UrlHelper::getSiteUrl($path, null, null, $row['locale']);
                    if (craft()->config->get('addTrailingSlashesToUrls')) {
                        $url = rtrim($url, '/') . '/';
                    }
                    $unsortedLocalizedUrls[$row['locale']] = $url;
                }

                $locales = craft()->i18n->getSiteLocales();
                foreach ($locales as $locale) {
                    $localeId = $locale->getId();
                    if (isset($unsortedLocalizedUris[$localeId])) {
                        $urlParts = parse_url($unsortedLocalizedUris[$localeId]);

                        array_push($localizedUris, "/" . $urlParts['path']);
                    }
                }
            } else {
                array_push($localizedUris, "/" . $element->uri);
            }
        }

        return $localizedUris;
    }

    /**
     * @return  string The name of the plugin
     */
    public function getPluginName()
    {
        $retourPlugin = craft()->plugins->getPlugin('retour');
        $result = $retourPlugin->getName();

        return $result;
    }

    /**
     * @return  mixed Returns the list of matching schemes
     */
    public function getMatchesList()
    {
        $result = array(
            'exactmatch' => Craft::t('Exact Match'),
            'regexmatch' => Craft::t('RegEx Match'),
        );

        // Add any plugins that offer the retourMatch() method
        foreach (craft()->plugins->getPlugins() as $plugin) {
            if (method_exists($plugin, "retourMatch")) {
                $result[$plugin->getClassHandle()] = $plugin->getName() . Craft::t(" Match");
            }
        }

        return $result;
    }
}
